### DISCUSSION TITLE
uMS

### DISCUSSION THREAD

**Discussion by yisar** [frejs/fre#209] [Ideas]
Trying to find a way to control the rendering process and scheduler externally.

uMS is a good source of inspiration, but not necessarily the answer.

https://github.com/reactjs/rfcs/pull/147

**Comment by fxxjdedd** [comment]
What does tearing means in react?

**Reply by yisar** [reply]
Because rendering is asynchronous, when a component depends on an external state, the external state changes during the rendering process, which will cause the first half of the component to be in the old state, and the next part to be rendered in the new state.

![61502196-ce109200-aa0d-11e9-9efc-6203545d367c](https://user-images.githubusercontent.com/12951461/103498034-5abb4400-4e7e-11eb-8047-47f8e4bc19e1.gif)

**Comment by yisar** [comment]
```js
const mutation = createMutation({
  snapshot: () => store.getValue(),
  verison: () => store.version,
  subscription: callback => {
    const subscription = store.subscribe(callback)
    return () => subscription.unsubscribe()
  }
})

function Example(){
  const snapshot = useMutation(mutation)
  ...
}
```

**Comment by yisar** [comment]
It is worth mentioning that tearing is only a problem. Our goal is not to solve this single problem, but to find a unified solution to control rendering through external state(mutation).

It's not just about rerendering, it should also support abandoning rendering.
